src/qililab/instruments/qblox/qblox_module.py

```diff
@@ -88,13 +80,14 @@ def __post_init__(self):
     settings: QbloxModuleSettings
     device: Pulsar | QcmQrm
     # Cache containing the last compiled pulse schedule for each sequencer
-    _cache: dict[int, PulseBusSchedule] = {}
+    cache: dict[int, PulseBusSchedule] = {}
 
     def __init__(self, settings: dict):
         # The sequences dictionary contains all the compiled sequences for each sequencer and a flag indicating whether
         # the sequence has been uploaded or not
-        self.sequences: dict[int, tuple[QpySequence, bool]] = {}  # {sequencer_idx: (program, True), ...}
+        self.sequences: dict[int, QpySequence] = {}  # {sequencer_idx: (program, True), ...}
```

Comment:
sequences and cache are now handled at the compiler. From there we can check if a given sequence or pulse_bus_schedule is to be reuploaded or not, so there is no need to have the boolean flag.
For repeating sequences there are different scenarios considered, which are different for the qrm and qcm since the qcm always has a 1 to one bus-sequencer correspondence but the qrm can have a single bus pointing to different sequencers.

Consider the second iteration of a software loop i.e. a circuit has been run and executed, sequences are saved to memory (qblox_module.cache and qblox_module.sequences)
case QCM:
1. If a sequence is uploaded to a new bus, don't do anything new (compile sequence and upload to bus)
2. If a sequence is uploaded to a previously used bus, overwrite the previous sequence (qblox_module.sequences[seq_idx] = program) - this is equivalent to 1
3. If nothing is to be compiled to a previously cached bus / sequencer. The cache for this bus will still contain the previously run sequences but nothing will happen since the execution at platform will only call `run` on the buses involved in the current list of `pulse_bus_schedules`
case QRM:
The only difference is the behaviour at 3. Because the QRM belongs to a single bus, we have to check that the pulse_bus_schedules in the circuit to be run are the same as the ones in cache

3. If the same sequencer is run in consecutive iterations, the first sequence will be overwritten by the new one like in 2. At a middle point in execution (inside the `compiler` method), `qblox_module.sequences` and `qblox_module.cache` will contain the union of sequencers to run in the current execution and sequencers that run in the previous one but are not to run in the current one. Then we remove the ones with sequencers not found in the `pulse_bus_schedules` belonging to the current circuit, see lines
```python
        # check for sequences which where not in the PulseSchedule but are uploaded and erase them from cache
        # this is only needed for the qrm since it has multiple sequencers for the same bus
        missing_seq_ids = [
            cached_seq_id
            for cached_seq_id in qrm.cache.keys()
            if cached_seq_id not in (sequencer.identifier for sequencer, _ in sequencer_qrm_bus_schedules)
        ]
        # pop from cache. qblox_module takes care of popping from sequences
        for seq_id in missing_seq_ids:
            _ = qrm.cache.pop(seq_id)
            _ = qrm.sequences.pop(seq_id)
```
in the compiler